src/main/java/com/google/step/snippet/servlets/SearchServlet.java

```diff
@@ -97,13 +103,53 @@ public void doGet(HttpServletRequest request, HttpServletResponse response)
     }
     String query = encodeValue(param);
     List<Card> allCards = new ArrayList<>();
-    for (Client client : clients) {
-      String link = getLink(client.getCseId(), query);
-      Card card = client.search(link);
-      if (card != null) {
-        allCards.add(card);
-      }
+    ExecutorService executor = Executors.newFixedThreadPool(THREAD_COUNT);
+    List<Callable<Card>> cardCallbacks =
+        Arrays.asList(
+            () -> {
+              String w3Link = getLink(w3Client.getCseId(), query);
+              if (w3Link != null) {
+                return w3Client.search(w3Link);
+              }
+              return null;
+            },
+            () -> {
+              String stackLink = getLink(stackClient.getCseId(), query);
+              if (stackLink != null) {
+                return stackClient.search(stackLink);
+              }
+              return null;
+            },
+            () -> {
+              String geeksLink = getLink(geeksClient.getCseId(), query);
+              if (geeksLink != null) {
+                return geeksClient.search(geeksLink);
+              }
+              return null;
+            });
+    /* thread process completion takes roughly 1.5 to 2.5 seconds, thus a timeout of
+     * 3 seconds was chosen.
+     */
+    try {
+      executor.invokeAll(cardCallbacks, 3L, TimeUnit.SECONDS).stream()
```

Comment:
I would base the timeout on what would be reasonable for a user to wait for results, not how long it's necessary to get results on your machine. (Remember it might take a different amount of time on someone else's machine, or dependent on their internet connection! Although it's probably (?) faster on a server since it has ethernet.)

Reply:
What would be the best way to go about estimating this? "What's reasonable" seems like a pretty wide range.

Reply:
Fair. There's HCI studies that discuss how long a user is willing to wait, but most of the studies boil down to "faster is better so just be as fast as possible". 3 is fine I suppose, as long as it usually enough on your computer.